src/exchange.ts

```diff
@@ -33,68 +34,80 @@ export const devtoolsExchange: Exchange = ({ client, forward }) => {
     const handler = messageHandlers[e.detail.type];
     handler && handler(client)(e.detail);
   });
-  sendToContentScript({ type: 'init' });
 
-  return (ops$) => {
-    const sharedOps$ = pipe(ops$, map(addOperationContext), share);
+  // Tell the content script we are present
+  sendToContentScript({ type: 'init' });
 
-    const isDevtoolsOp = (o: Operation) =>
-      Boolean(o.context.meta && o.context.meta.source === 'Devtools');
+  // Forward debug events to content script
+  client.debugTarget!.addEventListener((event) =>
+    sendToContentScript({
+      type: 'debug',
+      data: event,
+    })
+  );
 
-    const appOps$ = pipe(
-      sharedOps$,
-      filter((o) => !isDevtoolsOp(o)),
-      tap(handleOperation),
-      forward,
-      map(addOperationResponseContext),
-      tap(handleOperation)
-    );
+  return (ops$) => pipe(ops$, tap(handleOperation), forward, tap(handleResult));
+};
 
-    const devtoolsOps$ = pipe(
-      sharedOps$,
-      filter((o) => isDevtoolsOp(o)),
-      forward
-    );
+/** Handle outgoing operations */
+const handleOperation = (operation: Operation) => {
+  if (operation.operationName === 'teardown') {
+    return sendDebugToContentScript({
+      type: 'teardown',
+      message: 'The operation has been torn down',
+      operation,
+    });
+  }
 
-    return merge([appOps$, devtoolsOps$]);
-  };
+  return sendDebugToContentScript({
+    type: 'execution',
+    message: 'The client has recieved an execute command.',
```

Comment:
nit: small typo. Also from a `message` standpoint, should they be more user friendly? The teardown and execute could be just described as “has started/cancelled an operation”

Reply:
I do think we should clean this up so suggestions are really appreciated!

I'm trying to avoid wording that might be misinterpreted ('started' can happen multiple times hence why we've avoided that term and 'cancelled' also has certain implications).

The other problem we have is the term _operation_ which @wgolledge, @sofiapoh and I have found challenging to communicate. If you look at the rows below, each could be considered an _operation_ in the sense that all events are for the given query+variables combination and therefore that row has a unique key - but that would be incorrect. 

![Screenshot 2020-03-27 at 09 55 42](https://user-images.githubusercontent.com/10779424/77744224-37bb5400-7011-11ea-8a8a-235d8cee2722.png)

An operation is not uniquely identified by it's key. Here's an example of two _operations_ which have the same key, but are not the same operation.

```js
// Operation 1
{
  key: 1234,
  query: `
    query UserIds {
      users {
        id
      }
    }
  `,
  context: {
    requestPolicy: 'cache-only',
  }
}

// Operation 2 (same key, not the same operation)
{
  key: 1234,
  query: `
    query UserIds {
      users {
        id
      }
    }
  `,
  context: {
    requestPolicy: 'cache-first',
  }
}
```

So for a teardown, I think what we're working with is not the teardown of an operation but instead the teardown of an _operation stream_.

Reply:
Describing them as `operation streams` sounds great to me. As long as the user is aware that an overarching _operation_ can be comprised of multiple streams.

Reply:
> overarching operation can be comprised of multiple streams

Do you mean this?

> overarching stream can be comprised of multiple operations

Damn this whole wording thing is confusing 😕

Reply:
I don’t think it’s a good idea to operations with identical keys and differing context options as separate.
They’ll travel through the exchange pipeline as mostly identical.
When you have a dedup exchange they wouldn’t even be active at the same time.
And when a request is sent to the server they’ll behave identically.

Since we display debug events as part of an “operations history and activity” it makes sense in my opinion to treat them as identical and simply indicate when they’ve been dispatched, which can have happened multiple times, and which due to reexecution can happen multiple times anyway.

If we treat them as separate then component updates, cache invalidation/updates, and result timing may become confusing 😅

Reply:
I agree with what you're saying, and I wish we could, but we can't use the same term for two distinctly different things (an operation stream, and an operation).

We're exposing implementation details in devtools, if someone calls _client.executeRequestOperation_ twice with two different operations (differing in context for example), we have to be able to present them as two different entities - calling them both the same operation wouldn't work.

Also in that screenshot earlier, if the operation stream is referred to as the _operation_, what do we refer to the entity that was passed to _executeRequestOperation_? Given that we're calling it an operation in the code, I think we should stick to that.

Reply:
This seems to be the most accurate and least confusing way to describe what our flow is:

 - Operations can be used to setup and teardown operation streams
 - Operations can also be used to trigger refetches on an operation stream
 - A successful/failed operation will result in an operation response on the respective stream